lib/sources.nix

```diff
@@ -26,12 +26,28 @@ rec {
     (type == "symlink" && lib.hasPrefix "result" baseName)
   );
 
-  cleanSource = builtins.filterSource cleanSourceFilter;
+  cleanSource = filterSource cleanSourceFilter;
+
+  # Like `builtins.filterSource`, except it will compose with itself,
+  # allowing you to chain multiple calls together without any
+  # intermediate copies being put in the nix store.
+  filterSource = f: src:
+    let isPath = builtins.typeOf src == "path";
+    in assert isLocalSource src; rec {
```

Comment:
Instead of asserting here, why don't we just do `builtins.storePath src` if it's in the store already? You can't filter in that case but it's already in the store so there's no need anyway.

Reply:
Or maybe not even do `storePath`, just pass `src` as-is.

Reply:
I don't think it's a good idea to allow filterSource to accept inputs that it won't filter. That would be very misleading

Reply:
Filtering is an optimization, though...

Reply:
I don't think so. Filtering can be semantically relevant. e.g. If they want to filter out a particular source file.

Reply:
Maybe this should take its arguments in an attrset so that we can extend it with new arguments (like, `fallBackNoFilter`) in the future?

Reply:
👍